File: web/js/utils/ws-chunk.js
//     ▄▄ ▄▄ ▄▄  ▄▄▄▄  ▄▄▄▄ ▄▄    ▄▄▄▄▄ ▄▄▄▄
//     ██ ██ ██ ██ ▄▄ ██ ▄▄ ██    ██▄▄  ██▄█▄   Copyright (c) 2026 Julian Storer
//   ▄▄█▀ ▀███▀ ▀███▀ ▀███▀ ██▄▄▄ ██▄▄▄ ██ ██   AGPL-3.0-or-later - see LICENSE

/**
 * Reassembly for messages the server had to split.
 *
 * A WebSocket message has no size limit on the wire, but every client library
 * imposes one on what it will accept. Node's built-in WebSocket — what the
 * headless engine host runs on — refuses anything over 128 MiB by closing the
 * connection, which nothing recovers from on its own: the client reconnects,
 * the server re-sends the same message, and the conversation stays unusable.
 * So the server splits a message that large into binary frames (see
 * cmd/juggler/server/ws_chunk.go) and this puts it back together.
 *
 * The frame layout is fixed on both sides:
 *
 *   magic   4 bytes   "JGC1"
 *   kind    1 byte    what the reassembled bytes are (WS_CHUNK_KIND_*)
 *   id      8 bytes   uint64 BE, unique per connection
 *   index   4 bytes   uint32 BE
 *   total   4 bytes   uint32 BE
 *   data    remainder
 * @module utils/ws-chunk
 */

/** Bytes of fixed header before a chunk's data. Matches wsChunkHeaderSize. */
export const WS_CHUNK_HEADER_SIZE = 21;

/** Reassembled bytes are a UTF-8 text message. */
export const WS_CHUNK_KIND_TEXT = 0;

/** Reassembled bytes are themselves a binary message. */
export const WS_CHUNK_KIND_BINARY = 1;

const MAGIC = [0x4a, 0x47, 0x43, 0x31]; // "JGC1"

/**
 * @typedef {object} WSChunkFrame
 * @property {number} kind - One of WS_CHUNK_KIND_*
 * @property {string} id - The run this frame belongs to
 * @property {number} index - Position within the run, from 0
 * @property {number} total - How many frames the run has
 * @property {Uint8Array} data - This frame's slice of the payload
 */

/**
 * Read a chunk frame. Returns null for anything that is not one, so an ordinary
 * binary message is never mistaken for a chunk and swallowed.
 * @param {ArrayBuffer|Uint8Array} raw - A received binary message
 * @returns {WSChunkFrame|null} The parsed frame, or null if it is not one
 */
export function parseWSChunkFrame(raw) {
  const bytes = raw instanceof Uint8Array ? raw : new Uint8Array(raw);
  if (bytes.length < WS_CHUNK_HEADER_SIZE) return null;
  for (let i = 0; i < MAGIC.length; i++) {
    if (bytes[i] !== MAGIC[i]) return null;
  }

  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  // The id is read as two 32-bit halves rather than a BigInt purely to keep the
  // key a plain string; nothing here does arithmetic on it.
  const id = `${view.getUint32(5)}-${view.getUint32(9)}`;
  const index = view.getUint32(13);
  const total = view.getUint32(17);
  if (total === 0 || index >= total) return null;

  return {
    kind: view.getUint8(4),
    id,
    index,
    total,
    data: bytes.subarray(WS_CHUNK_HEADER_SIZE)
  };
}

/**
 * Collects the frames of a split message until it is whole again.
 *
 * One of these belongs to one connection. The server writes a run of frames
 * from a single goroutine, so in practice they arrive contiguous and in order,
 * but nothing here assumes that: frames are placed by index and the run
 * completes when every slot is filled. The id is what keeps two runs apart, so
 * a dropped frame strands its own run rather than splicing the next one onto
 * it.
 */
export class WSChunkReassembler {
  constructor() {
    /** @type {Map<string, {total: number, kind: number, parts: (Uint8Array|undefined)[], received: number, bytes: number}>} @private */
    this._runs = new Map();
  }

  /**
   * Offer a received binary message.
   * @param {ArrayBuffer|Uint8Array} raw - The received binary message
   * @returns {{kind: number, bytes: Uint8Array}|null} The reassembled message
   *   once its last frame has arrived; null while the run is still incomplete,
   *   or if `raw` was not a chunk frame at all (distinguish with
   *   {@link parseWSChunkFrame} when that matters)
   */
  accept(raw) {
    const frame = parseWSChunkFrame(raw);
    if (!frame) return null;

    let run = this._runs.get(frame.id);
    if (!run) {
      run = {
        total: frame.total,
        kind: frame.kind,
        parts: new Array(frame.total),
        received: 0,
        bytes: 0
      };
      this._runs.set(frame.id, run);
    }
    // A run whose frames disagree about their own shape is not a run. Drop it
    // rather than assembling something that was never sent.
    if (run.total !== frame.total || run.kind !== frame.kind) {
      this._runs.delete(frame.id);
      return null;
    }
    // First writer wins, so a retransmitted frame cannot double-count and leave
    // the run permanently one short of complete.
    if (run.parts[frame.index] === undefined) {
      run.parts[frame.index] = frame.data;
      run.received++;
      run.bytes += frame.data.length;
    }
    if (run.received !== run.total) return null;

    this._runs.delete(frame.id);
    const bytes = new Uint8Array(run.bytes);
    let at = 0;
    for (const part of run.parts) {
      if (!part) continue;
      bytes.set(part, at);
      at += part.length;
    }
    return { kind: run.kind, bytes };
  }

  /**
   * Forget every partial run. Called when the link dies: the frames still
   * outstanding belong to a connection that will never finish sending them, and
   * the next connection starts its ids again from one.
   */
  reset() {
    this._runs.clear();
  }

  /** @returns {number} How many runs are part-way through. Diagnostics only. */
  get pending() {
    return this._runs.size;
  }
}

File: cmd/juggler/server/ws_chunk.go
//     ▄▄ ▄▄ ▄▄  ▄▄▄▄  ▄▄▄▄ ▄▄    ▄▄▄▄▄ ▄▄▄▄
//     ██ ██ ██ ██ ▄▄ ██ ▄▄ ██    ██▄▄  ██▄█▄   Copyright (c) 2026 Julian Storer
//   ▄▄█▀ ▀███▀ ▀███▀ ▀███▀ ██▄▄▄ ██▄▄▄ ██ ██   AGPL-3.0-or-later - see LICENSE

package server

import (
	"encoding/binary"
	"time"

	"github.com/gorilla/websocket"
)

// A WebSocket message has no size limit on the wire, but every client library
// imposes one on what it will accept, and the server has no way to discover it.
// Node's built-in WebSocket (undici, which is what the headless engine host
// runs on) defaults to 128 MiB and answers anything larger with a close 1009
// before it has read a single byte of the payload. Nothing recovers from that
// on its own: the client reconnects, the server re-sends the same message, and
// the conversation is unusable for as long as it stays that size.
//
// So a message that is too big for any plausible client is split here, and the
// client puts it back together. The frames are binary because the payload being
// carried is UTF-8 text: slicing text into JSON string fields would mean
// respecting rune boundaries and re-escaping every slice — a second full pass
// over a payload whose size is the entire problem — where raw bytes need
// neither, and the client decodes UTF-8 once at the end.
//
//	magic   4 bytes   "JGC1"
//	kind    1 byte    what the reassembled bytes are (see wsChunkKind*)
//	id      8 bytes   uint64 BE, unique per connection
//	index   4 bytes   uint32 BE
//	total   4 bytes   uint32 BE
//	data    remainder
//
// Concatenating every frame's data in index order reproduces the original
// payload byte for byte.
const (
	wsChunkHeaderSize = 21

	// wsChunkKindText marks reassembled bytes that are a UTF-8 text message —
	// what an unchunked send would have written as websocket.TextMessage.
	wsChunkKindText byte = 0

	// wsChunkKindBinary marks reassembled bytes that are themselves a binary
	// message and must not be decoded as text.
	wsChunkKindBinary byte = 1

	// wsChunkThreshold is the largest message written whole. Above it the
	// message is split into pieces of this size.
	//
	// Ordinary traffic is nowhere near it — sync deltas, status, tool output
	// chunks are kilobytes — so nothing on the hot path pays for this at all.
	// What it has to clear is the full-state resync of a large conversation,
	// which runs to megabytes and has no upper bound but the document's size.
	//
	// 8 MiB is chosen from both ends: far enough above ordinary traffic that
	// chunking stays an exceptional path worth logging when it happens, and far
	// enough below the smallest client ceiling known (undici's 128 MiB) that a
	// client with a stricter limit still has an order of magnitude of headroom.
	// It also keeps the frame count sane — a 100 MB document is 13 frames, not
	// the thousands a 16 KiB chunker would produce.
	wsChunkThreshold = 8 * 1024 * 1024
)

var wsChunkMagic = [4]byte{'J', 'G', 'C', '1'}

// wsChunkHeader is the fixed part of a chunk frame.
type wsChunkHeader struct {
	kind  byte
	id    uint64
	index uint32
	total uint32
}

// encodeWSChunkHeader writes h into dst, which must be at least
// wsChunkHeaderSize long.
func encodeWSChunkHeader(dst []byte, h wsChunkHeader) {
	copy(dst[0:4], wsChunkMagic[:])
	dst[4] = h.kind
	binary.BigEndian.PutUint64(dst[5:13], h.id)
	binary.BigEndian.PutUint32(dst[13:17], h.index)
	binary.BigEndian.PutUint32(dst[17:21], h.total)
}

// parseWSChunkHeader splits a frame into its header and data. Reports false for
// anything that is not a well-formed chunk frame, so an ordinary binary message
// is never mistaken for one.
func parseWSChunkHeader(frame []byte) (wsChunkHeader, []byte, bool) {
	if len(frame) < wsChunkHeaderSize {
		return wsChunkHeader{}, nil, false
	}
	if string(frame[0:4]) != string(wsChunkMagic[:]) {
		return wsChunkHeader{}, nil, false
	}
	h := wsChunkHeader{
		kind:  frame[4],
		id:    binary.BigEndian.Uint64(frame[5:13]),
		index: binary.BigEndian.Uint32(frame[13:17]),
		total: binary.BigEndian.Uint32(frame[17:21]),
	}
	if h.total == 0 || h.index >= h.total {
		return wsChunkHeader{}, nil, false
	}
	return h, frame[wsChunkHeaderSize:], true
}

// wsChunkCount is how many frames a payload of this size splits into.
func wsChunkCount(payloadLen int) int {
	if payloadLen <= 0 {
		return 1
	}
	return (payloadLen + wsChunkThreshold - 1) / wsChunkThreshold
}

// writeChunked writes one oversized payload as a run of chunk frames.
//
// Only writePump calls it (through writeOne), so the one-writer-per-connection
// rule gorilla requires still holds, and the run is contiguous: no other
// message can interleave with it and the client can rely on the frames arriving
// in order. The id rides along regardless so that if that ever stops being true
// the client can detect it rather than silently splice two messages together.
//
// Each frame is written through NextWriter and two Writes — the header, then a
// slice of the caller's payload — so nothing the size of the payload is copied
// or allocated on the way out. The write deadline is armed per frame rather
// than once for the whole run: a run is exactly as many independent writes as
// it has frames, and holding one deadline across all of them would give a large
// document less time per byte the larger it got.
func (c *WSClient) writeChunked(payload []byte, kind byte) error {
	c.chunkSeq++
	id := c.chunkSeq
	total := wsChunkCount(len(payload))

	var header [wsChunkHeaderSize]byte
	for index := 0; index < total; index++ {
		start := index * wsChunkThreshold
		end := start + wsChunkThreshold
		if end > len(payload) {
			end = len(payload)
		}
		encodeWSChunkHeader(header[:], wsChunkHeader{
			kind:  kind,
			id:    id,
			index: uint32(index),
			total: uint32(total),
		})

		_ = c.conn.SetWriteDeadline(time.Now().Add(wsWriteTimeout))
		w, err := c.conn.NextWriter(websocket.BinaryMessage)
		if err != nil {
			return err
		}
		if _, err = w.Write(header[:]); err != nil {
			_ = w.Close()
			return err
		}
		if _, err = w.Write(payload[start:end]); err != nil {
			_ = w.Close()
			return err
		}
		if err = w.Close(); err != nil {
			return err
		}
	}
	return nil
}
